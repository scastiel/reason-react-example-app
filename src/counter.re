open Interval;

let component = ReasonReact.statelessComponent "Counter";

let make
    counter::(counter: int)
    intervalId::(intervalId: option intervalId)
    increment::(increment: unit => unit)
    incrementEverySecond::(incrementEverySecond: unit => unit)
    stopIncrementingEverySecond::(stopIncrementingEverySecond: intervalId => unit)
    decrement::(decrement: unit => unit)
    _children => {
  ...component,
  render: fun self =>
    <div>
      (ReasonReact.stringToElement ("Counter: " ^ string_of_int counter))
      <button className="plus-button" onClick=(self.handle (fun _ _ => increment ()))>
        (ReasonReact.stringToElement "+")
      </button>
      (
        switch intervalId {
        | None =>
          <button
            className="plus-1s-button" onClick=(self.handle (fun _ _ => incrementEverySecond ()))>
            (ReasonReact.stringToElement "+ every 1s")
          </button>
        | Some intervalId =>
          <button
            className="stop-plus-1s-button"
            onClick=(self.handle (fun _ _ => stopIncrementingEverySecond intervalId))>
            (ReasonReact.stringToElement "stop +")
          </button>
        }
      )
      <button className="minus-button" onClick=(self.handle (fun _ _ => decrement ()))>
        (ReasonReact.stringToElement "-")
      </button>
    </div>
};

let counter =
  ReasonReact.wrapReasonForJs
    ::component
    (
      fun jsProps =>
        make
          counter::jsProps##counter
          intervalId::jsProps##intervalId
          increment::jsProps##increment
          incrementEverySecond::jsProps##incrementEverySecond
          stopIncrementingEverySecond::jsProps##stopIncrementingEverySecond
          decrement::jsProps##decrement
          [||]
    );
