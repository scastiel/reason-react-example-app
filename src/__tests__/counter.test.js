import React from 'react'
import { counter as Counter } from '../counter.re'
import { shallow } from 'enzyme'

describe('Counter component', () => {
  it('renders with value 0 without intervalId', () => {
    const wrapper = shallow(<Counter counter={0} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with value 10 without intervalId', () => {
    const wrapper = shallow(<Counter counter={10} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with value 10 with intervalId', () => {
    const wrapper = shallow(<Counter counter={10} intervalId={1234} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('calls increment when plus button is clicked', () => {
    const increment = jest.fn()
    const wrapper = shallow(
      <Counter counter={10} increment={increment} decrement={() => {}} />
    )
    wrapper.find('.plus-button').simulate('click')
    expect(increment.mock.calls.length).toEqual(1)
  })

  it('calls decrement when minus button is clicked', () => {
    const decrement = jest.fn()
    const wrapper = shallow(
      <Counter counter={10} increment={() => {}} decrement={decrement} />
    )
    wrapper.find('.minus-button').simulate('click')
    expect(decrement.mock.calls.length).toEqual(1)
  })

  it('calls incrementEverySecond when "+ every 1s" button is clicked', () => {
    const incrementEverySecond = jest.fn()
    const wrapper = shallow(
      <Counter counter={10} incrementEverySecond={incrementEverySecond} />
    )
    wrapper.find('.plus-1s-button').simulate('click')
    expect(incrementEverySecond.mock.calls.length).toEqual(1)
  })

  it('calls stopIncrementingEverySecond when "+ every 1s" button is clicked', () => {
    const stopIncrementingEverySecond = jest.fn()
    const wrapper = shallow(
      <Counter
        counter={10}
        intervalId={[1234]} // some 1234
        stopIncrementingEverySecond={stopIncrementingEverySecond}
      />
    )
    wrapper.find('.stop-plus-1s-button').simulate('click')
    expect(stopIncrementingEverySecond.mock.calls.length).toEqual(1)
    expect(stopIncrementingEverySecond.mock.calls[0][0]).toEqual(1234)
  })
})
