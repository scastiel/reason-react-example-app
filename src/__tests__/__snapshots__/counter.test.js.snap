// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component renders with value 0 without intervalId 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    Counter: 0
    <button
        className="plus-button"
        onClick={[Function]}
    >
        +
    </button>
    <button
        className="plus-1s-button"
        onClick={[Function]}
    >
        + every 1s
    </button>
    <button
        className="minus-button"
        onClick={[Function]}
    >
        -
    </button>
</div>,
  "nodes": Array [
    <div>
      Counter: 0
      <button
            className="plus-button"
            onClick={[Function]}
      >
            +
      </button>
      <button
            className="plus-1s-button"
            onClick={[Function]}
      >
            + every 1s
      </button>
      <button
            className="minus-button"
            onClick={[Function]}
      >
            -
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Counter
        counter={0}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance":  {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "enqueueMethod": [Function],
        "handleMethod": [Function],
        "props": Object {
          "counter": 0,
        },
        "refs": Object {},
        "self": [Function],
        "state": Object {
          "reasonState": 0,
          "reasonStateVersion": 1,
          "reasonStateVersionUsedToComputeSubelements": 1,
        },
        "transitionNextTotalState": [Function],
        "updateMethod": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          Counter: 0
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="plus-1s-button"
                    onClick={[Function]}
          >
                    + every 1s
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          Counter: 0
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="plus-1s-button"
                    onClick={[Function]}
          >
                    + every 1s
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Counter
    counter={0}
/>,
}
`;

exports[`Counter component renders with value 10 with intervalId 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    Counter: 10
    <button
        className="plus-button"
        onClick={[Function]}
    >
        +
    </button>
    <button
        className="stop-plus-1s-button"
        onClick={[Function]}
    >
        stop +
    </button>
    <button
        className="minus-button"
        onClick={[Function]}
    >
        -
    </button>
</div>,
  "nodes": Array [
    <div>
      Counter: 10
      <button
            className="plus-button"
            onClick={[Function]}
      >
            +
      </button>
      <button
            className="stop-plus-1s-button"
            onClick={[Function]}
      >
            stop +
      </button>
      <button
            className="minus-button"
            onClick={[Function]}
      >
            -
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Counter
        counter={10}
        intervalId={1234}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance":  {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "enqueueMethod": [Function],
        "handleMethod": [Function],
        "props": Object {
          "counter": 10,
          "intervalId": 1234,
        },
        "refs": Object {},
        "self": [Function],
        "state": Object {
          "reasonState": 0,
          "reasonStateVersion": 1,
          "reasonStateVersionUsedToComputeSubelements": 1,
        },
        "transitionNextTotalState": [Function],
        "updateMethod": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          Counter: 10
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="stop-plus-1s-button"
                    onClick={[Function]}
          >
                    stop +
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
        "_debugID": 6,
        "_renderedOutput": <div>
          Counter: 10
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="stop-plus-1s-button"
                    onClick={[Function]}
          >
                    stop +
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Counter
    counter={10}
    intervalId={1234}
/>,
}
`;

exports[`Counter component renders with value 10 without intervalId 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    Counter: 10
    <button
        className="plus-button"
        onClick={[Function]}
    >
        +
    </button>
    <button
        className="plus-1s-button"
        onClick={[Function]}
    >
        + every 1s
    </button>
    <button
        className="minus-button"
        onClick={[Function]}
    >
        -
    </button>
</div>,
  "nodes": Array [
    <div>
      Counter: 10
      <button
            className="plus-button"
            onClick={[Function]}
      >
            +
      </button>
      <button
            className="plus-1s-button"
            onClick={[Function]}
      >
            + every 1s
      </button>
      <button
            className="minus-button"
            onClick={[Function]}
      >
            -
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Counter
        counter={10}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance":  {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "enqueueMethod": [Function],
        "handleMethod": [Function],
        "props": Object {
          "counter": 10,
        },
        "refs": Object {},
        "self": [Function],
        "state": Object {
          "reasonState": 0,
          "reasonStateVersion": 1,
          "reasonStateVersionUsedToComputeSubelements": 1,
        },
        "transitionNextTotalState": [Function],
        "updateMethod": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          Counter: 10
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="plus-1s-button"
                    onClick={[Function]}
          >
                    + every 1s
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
        "_debugID": 4,
        "_renderedOutput": <div>
          Counter: 10
          <button
                    className="plus-button"
                    onClick={[Function]}
          >
                    +
          </button>
          <button
                    className="plus-1s-button"
                    onClick={[Function]}
          >
                    + every 1s
          </button>
          <button
                    className="minus-button"
                    onClick={[Function]}
          >
                    -
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Counter
    counter={10}
/>,
}
`;
